### DISCUSSION TITLE
How to disable "rsc" in searchParams?

### DISCUSSION THREAD

**Discussion by neila-a** [vercel/next.js#55011] [Help] [answered]
### Summary

I use ServiceWorker and CacheStorage for caching in my application. It has been working fine. But the URL provided by the Router of Next.js is always like this: "https://someone.com/...?rsc=somenumber...", and the number of rsc will change every time. This caused my ServiceWorker to cache each page twice, or more. What I want to ask is, what is the function of "rsc" in searchParam?

**Comment by devjiwonchoi** [comment]
I believe that `rsc` is for RSC Payload, related to caching and prefetching your route.
I'm assuming that you are experiencing this while using the `Link` component, you can opt this out by setting `prefetch: false`.

Or else, you have several options to opt out based on the need, and this is one of them.

### Opt out - Full Route Cache

- Using a [Dynamic Function](https://nextjs.org/docs/app/building-your-application/caching#dynamic-functions): This will opt the route out from the Full Route Cache and dynamically render it at request time. The Data Cache can still be used.
- Using the `dynamic = 'force-dynamic'` or `revalidate = 0` route segment config options: This will skip the Full Route Cache and the Data Cache. Meaning components will be rendered and data fetched on every incoming request to the server. The Router Cache will still apply as it's a client-side cache.
- Opting out of the [Data Cache](https://nextjs.org/docs/app/building-your-application/caching#data-cache): If a route has a fetch request that is not cached, this will opt the route out of the Full Route Cache. The data for the specific fetch request will be fetched for every incoming request. Other fetch requests that do not opt out of caching will still be cached in the Data Cache. This allows for a hybrid of cached and uncached data.

I'd recommend you to check out the `Opting out` sections of the following reference.

P.S. If none of these work, it would be great if you share a repro.

### Reference
- [Caching](https://nextjs.org/docs/app/building-your-application/caching)

**Reply by dpnova** [reply]
on click I see an _rsc request go out even when prefetch is false on links in client modules. Currently running with gutting the prefetch reducer with a package patch.

**Reply by dpnova** [reply, marked as answer]
fix here:
```
diff --git a/dist/client/components/router-reducer/reducers/prefetch-reducer.js b/dist/client/components/router-reducer/reducers/prefetch-reducer.js
index b76da873519d5e524497b5f72ecf0e545f595a20..f293a268f006bf864b75dbd59fc2c3b01e611a0b 100644
--- a/dist/client/components/router-reducer/reducers/prefetch-reducer.js
+++ b/dist/client/components/router-reducer/reducers/prefetch-reducer.js
@@ -26,17 +26,6 @@ const _prefetchcacheutils = require("../prefetch-cache-utils");
 const prefetchQueue = new _promisequeue.PromiseQueue(5);
 function prefetchReducer(state, action) {
     // let's prune the prefetch cache before we do anything else
-    (0, _prefetchcacheutils.prunePrefetchCache)(state.prefetchCache);
-    const { url } = action;
-    url.searchParams.delete(_approuterheaders.NEXT_RSC_UNION_QUERY);
-    (0, _prefetchcacheutils.getOrCreatePrefetchCacheEntry)({
-        url,
-        nextUrl: state.nextUrl,
-        prefetchCache: state.prefetchCache,
-        kind: action.kind,
-        tree: state.tree,
-        buildId: state.buildId
-    });
     return state;
 }
 ```

**Reply by neila-a** [reply]
I don't think changing the compiled code is an effective means, but this at least works.

**Reply by dpnova** [reply]
yeah, far from ideal... part of the point of sharing is to highlight how bad it is that you can't configure this :D

**Comment by neila-a** [comment]
You are right, but this is all about turning off the RSC feature, but how to disable and only disable rsc in the URL?

**Comment by arfa123** [comment]
I am developing a static web app with `output: "export"`, where all components and pages are client-based. However, I’ve noticed that route requests include a `?_rsc={unique string}` query parameter. My goal is to create a fully offline-supported static web app, and I am prefetching all assets and routes at the start of the app. Unfortunately, this `?_rsc=` query parameter on each route is causing issues and preventing the app from achieving full offline support.

**Reply by dpnova** [reply]
I just created this patch with pnpm-patch - yarn has a similar feature 

```
diff --git a/dist/client/components/router-reducer/reducers/prefetch-reducer.js b/dist/client/components/router-reducer/reducers/prefetch-reducer.js
index b76da873519d5e524497b5f72ecf0e545f595a20..f293a268f006bf864b75dbd59fc2c3b01e611a0b 100644
--- a/dist/client/components/router-reducer/reducers/prefetch-reducer.js
+++ b/dist/client/components/router-reducer/reducers/prefetch-reducer.js
@@ -26,17 +26,6 @@ const _prefetchcacheutils = require("../prefetch-cache-utils");
 const prefetchQueue = new _promisequeue.PromiseQueue(5);
 function prefetchReducer(state, action) {
     // let's prune the prefetch cache before we do anything else
-    (0, _prefetchcacheutils.prunePrefetchCache)(state.prefetchCache);
-    const { url } = action;
-    url.searchParams.delete(_approuterheaders.NEXT_RSC_UNION_QUERY);
-    (0, _prefetchcacheutils.getOrCreatePrefetchCacheEntry)({
-        url,
-        nextUrl: state.nextUrl,
-        prefetchCache: state.prefetchCache,
-        kind: action.kind,
-        tree: state.tree,
-        buildId: state.buildId
-    });
     return state;
 }
 ```

**Comment by khanakia** [comment]
Is there a workaround how to stop sending _rsc http://localhost:2305/apps/a1/posts/u0zz2euf2rfk76h58a/edit?_rsc=13zaz

I do not want the next.js call `rsc` request everytime I open the pages

`src/app/apps/[appid]/posts/[id]/edit/page.tsx`
```tsx
'use client'

export const dynamic = 'force-static'

import { EditPageView } from '@fasto/crud'

import EditPost from '@/features/sa/components/pages/posts/EditPost'

export default function Page() {
  return <div>Test</div>
}

```

**Reply by dpnova** [reply]
i have pasted a solution above

**Comment by achepukov** [comment]
pretty unfair game to change the api without knowing the customers
Can we move this url parameter to dedicated header?

**Comment by samueltesfayegari** [comment]
This is pain in the ... I just made my whole page a client side, no use of URL params 👎

**Reply by dpnova** [reply]
we're switching to vite - next is overkill for a pure client/local app

**Reply by otterDeveloper** [reply]
using React Router inside next could be an alternative

**Reply by neila-a** [reply]
> 我们正在切换到 Vite - 接下来是纯客户端/本地应用程序的矫枉过正

Yes, I am switching to Vite. However, the high memory load of SPA is a serious issue, and what I want is to use CacheStorage as a server, storing the webpage on the hard disk rather than in memory. I believe Nextjs is the best choice for this.
